tests/engine/test_state.py

```diff
@@ -676,3 +679,68 @@ def test_n_map_states():
 def test_init_with_falsey_value():
     state = Success(result={})
     assert state.result == {}
+
+
+def test_state_pickle():
+    @dataclass
+    class Data:
+        bar: str
+        foo: int = 1
+
+    state = State(message="message", result=Data(bar="bar"))
+    new_state = cloudpickle.loads(cloudpickle.dumps(state))
+
+    assert state.message == new_state.message  # message not included in __eq__
+    assert state == new_state
+
+
+def test_state_pickle_with_unpicklable_result_raises():
```

Comment:
test name here is misleading - i think you mean 
```suggestion
def test_state_pickle_with_unpicklable_attribute_raises():
```

Reply:
Hm. I'm specifically referring to the `State.result` here though not any ole attribute

Reply:
But your PR prevents the raise when the unpickleable thing is `State.result` and this tests that the raise does occur for `State.x`?

Reply:
I guess here I could actually just set the result to `RLock()` instead of using a dataclass which simplifies the test

Reply:
I'll update the test to be simpler; not sure why I thought I needed to make it a class.

This test covers `State.result.x` being unpickable; really just needs to cover `State.result` being unpickable and not an `Exception` type

Reply:
Good catch :)

Reply:
https://github.com/PrefectHQ/prefect/pull/4605/commits/478cc1abeeb1933e83f1d7571b4d84644a1779c5